text/0000-react-global-hooks.md

```diff
@@ -0,0 +1,135 @@
+
+## Motivation
+
+TODO
+
+Inspiration:
+
+- [Zustand](https://github.com/pmndrs/zustand)
+- [Redux (in React)](https://react-redux.js.org/)
+- [Recoil](https://recoiljs.org/)
+
+## Detailed design
+
+**Quick overview:**
+```js
+import { createHooks } from '@xstate/react';
+import { authMachine } from './authMachine';
+import { interpret } from 'xstate';
+
+// Interpret the machine first (full interpretation control)
+const authService = interpret(authMachine, {/* ... */}).start();
+
+// Create a "hooks object" from the actor
+const auth = createHooks(authService);
+
+// ...
+
+const SomeForm = () => {
+  // Get full state snapshot and send() function
+  const [state, send] = auth.useActor();
+
+  // Get selected state
+  const isLoggedIn = auth.useSelector(state => state.hasTag('authorized'));
+
+  // ...
+}
+```
+
+**Details**
+```js
+import { createHooks } from '@xstate/react';
+import { authMachine } from './authMachine';
+
+const authService = interpret(authMachine, {/* ... */}).start();
+
+const auth = createHooks(authService);
+
+// Returned object has the actor ref instance
+auth.actorRef === authService; // true
+
+// Object has two extra methods to be used as hooks:
+
+// Hook for accessing [state, send] from actor
+const [state, send] = auth.useActor();
+
+// Hook for accessing selected state from actor
+const derivedState = auth.useSelector(selectorFn, comparator);
+```
+
+
+### Technical Background
+
+This builds on two existing hooks: `useActor()` and `useSelector()`.
+
+TODO
+
+### Implementation
+
+```ts
+// Simplified
+import { useActor, useSelector } from '@xstate/react';
+
+function createHooks(actorRef: ActorRef) {
+  return {
+    actorRef,
+    useActor() {
+      return useActor(actorRef);
+    },
+    useSelector(selector, comparator) {
+      return useSelector(actorRef, comparator);
+    }
+  };
+}
+```
+
+## How we teach this
+
+TODO
+
+## Drawbacks
+
+- There are fast-refresh/HMR compatibility concerns, similar to those listed [here (Zustand)](https://github.com/pmndrs/zustand/issues/908). However, we can probably still have "good enough" support with the same fast-refresh capabilities as currently exist:
+
+  > this kind of Fast Refresh support isnt that bad - kinda what we are doing at the moment anyway (just blowing up the old state and restarting)
+  > @Andarist (https://discord.com/channels/795785288994652170/976863583336022127/976875867655524412)
+
+- This may feel unfamiliar to React devs who would naturally reach for [React Context](https://reactjs.org/docs/context.html) for shared state; however, Zustand has already popularized the notion of "contextual state" outside of React Context: https://github.com/pmndrs/zustand#why-zustand-over-context
```

Comment:
Zustand is a little bit simpler because they don't create any side-effects at any kind of initialization time. The problem with `.start()` is that it has to be called **somewhere** and that it can't be called during SSR. At the very least we should provide guidance around this and some examples of how one *should* deal with this in the SSR context.